src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,223 @@
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _innerHandler = innerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response;
+            uint redirectCount = 0;
+            bool useCredentialCache = false;
+            while (true)
+            {
+                response = await SendRequestAsync(request, useCredentialCache, cancellationToken).ConfigureAwait(false);
+
+                if (!RequestNeedsRedirect(response))
+                {
+                    break;
+                }
+
+                // Clear the authorization header, if the request requires redirect.
+                request.Headers.Authorization = null;
+
+                // Just as with WinHttpHandler and CurlHandler, for security reasons, we drop the server credential if it is
+                // anything other than a CredentialCache. We allow credentials in a CredentialCache since they
+                // are specifically tied to URIs.
+                useCredentialCache = true;
+
+                Uri location = response.Headers.Location;
+                if (location == null)
+                {
+                    // No location header. Nothing to redirect to.
+                    break;
+                }
+
+                if (!location.IsAbsoluteUri)
+                {
+                    location = new Uri(request.RequestUri, location);
+                }
+
+                // Disallow automatic redirection from secure to non-secure schemes
+                bool allowed =
+                    (HttpUtilities.IsSupportedNonSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedScheme(location.Scheme)) ||
+                    (HttpUtilities.IsSupportedSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedSecureScheme(location.Scheme));
+                if (!allowed)
+                {
+                    break;
+                }
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Set up for the automatic redirect
+                request.RequestUri = location;
+
+                if (RequestRequiresForceGet(response.StatusCode, request.Method))
+                {
+                    request.Method = HttpMethod.Get;
+                    request.Content = null;
+                }
+
+                // Do the redirect.
+                response.Dispose();
+            }
+
+            return response;
+        }
+
+        private bool RequestNeedsRedirect(HttpResponseMessage response)
+        {
+            // Return if redirect not allowed.
```

Comment:
>redirect not allowed [](start = 25, length = 20)

Nit: it's not that the redirect isn't allowed per se, it's that the developer has asked us not to follow redirects automatically.  "not allowed" suggests, for example, a security restriction that forces us not to follow a redirect.

Reply:
Ok